File: packages/sdk/python/mcp-server/src/mcp_obs_server/oauth_validator.py
"""
OAuth token validation utilities for MCP servers
Uses HTTP request to mcp-obs platform for token validation
"""
from typing import Optional, List
import time
import httpx
from loguru import logger
from urllib.parse import urlencode

from .types import OAuthConfig, AuthContext


class OAuthTokenValidator:
    """OAuth token validator that validates tokens against mcp-obs platform"""

    def __init__(self, config: OAuthConfig):
        self.config = config
        self._http_client = httpx.AsyncClient(timeout=30.0)

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self._http_client.aclose()

    async def validate_token(self, token: str) -> Optional[AuthContext]:
        """
        Validate a Bearer token using HTTP request to mcp-obs platform

        Args:
            token: The Bearer token to validate

        Returns:
            AuthContext if valid, None if invalid
        """
        try:
            platform_url = (
                self.config.platform_url or
                f"https://{self.config.server_slug}.mcp-obs.com"
            )

            if self.config.debug:
                logger.debug(f"[OAuth] Validating token with platform: {platform_url}")

            # Make HTTP request to introspect token
            introspect_url = f"{platform_url}/api/mcp-oauth/introspect"

            form_data = {
                "token": token,
                "server_slug": self.config.server_slug
            }

            response = await self._http_client.post(
                introspect_url,
                headers={
                    "Content-Type": "application/x-www-form-urlencoded",
                },
                content=urlencode(form_data)
            )

            if not response.is_success:
                if self.config.debug:
                    logger.error(f"[OAuth] HTTP error {response.status_code}: {response.text}")
                return None

            introspection = response.json()

            if not introspection.get("active", False):
                if self.config.debug:
                    logger.error("[OAuth] Token is not active")
                return None

            # Build AuthContext from introspection response
            user_id = (
                introspection.get("sub") or
                introspection.get("mcp:user_id") or
                "unknown"
            )

            email = introspection.get("username", "unknown")
            name = introspection.get("name")
            image = introspection.get("picture")

            scopes = []
            if introspection.get("scope"):
                scopes = introspection["scope"].split(" ")

            client_id = introspection.get("client_id", "unknown")

            # Handle expiration time
            expires_at = int(time.time() * 1000) + 3600000  # Default: 1 hour from now
            if introspection.get("exp"):
                expires_at = int(introspection["exp"]) * 1000

            auth_context = AuthContext(
                user_id=user_id,
                email=email,
                name=name,
                image=image,
                scopes=scopes,
                client_id=client_id,
                expires_at=expires_at
            )

            if self.config.debug:
                logger.debug(f"[OAuth] Token validation successful for user: {auth_context.email}")

            return auth_context

        except Exception as error:
            logger.error(f"[OAuth] Token validation error: {error}")
            return None

    def extract_bearer_token(self, auth_header: Optional[str]) -> Optional[str]:
        """
        Extract Bearer token from Authorization header

        Args:
            auth_header: The Authorization header value

        Returns:
            The Bearer token or None if not found
        """
        if not auth_header or not auth_header.startswith("Bearer "):
            return None
        return auth_header[7:]  # Remove "Bearer " prefix

    def has_required_scopes(self, auth_context: AuthContext, required_scopes: List[str]) -> bool:
        """
        Check if token has required scopes

        Args:
            auth_context: The validated auth context
            required_scopes: Array of required scopes

        Returns:
            True if all required scopes are present
        """
        return all(scope in auth_context.scopes for scope in required_scopes)

    async def close(self):
        """Close the HTTP client"""
        await self._http_client.aclose()


# Convenience function for simple token validation
async def validate_token(config: OAuthConfig, token: str) -> Optional[AuthContext]:
    """
    Convenience function to validate a token without managing the validator instance

    Args:
        config: OAuth configuration
        token: Bearer token to validate

    Returns:
        AuthContext if valid, None if invalid
    """
    async with OAuthTokenValidator(config) as validator:
        return await validator.validate_token(token)

File: packages/server-sdk/src/oauth-validator.ts
/**
 * OAuth token validation utilities for MCP servers
 * Uses direct database validation (MCPlatform pattern)
 */

export interface TokenValidationResult {
  valid: boolean;
  token?: {
    id: string;
    accessToken: string;
    tokenType: string;
    scope?: string;
    expiresAt: Date;
    clientId: string;
    userId?: string;
    mcpServerId: string;
  };
  user?: {
    id: string;
    email: string;
    name?: string;
    image?: string;
  };
  client?: {
    id: string;
    clientId: string;
    clientName: string;
    clientUri?: string;
  };
  server?: {
    id: string;
    name: string;
    slug: string;
    organizationId: string;
    issuerUrl: string;
  };
  error?: string;
}

export interface OAuthConfig {
  /** MCP server slug (e.g., "test" for test.mcp-obs.com) */
  serverSlug: string;
  /** Platform base URL (e.g., "http://localhost:3000" or "https://my-org.mcp-obs.com") */
  platformUrl?: string;
  /** Enable debug logging */
  debug?: boolean;
}

export interface AuthContext {
  /** User ID from the OAuth token */
  userId: string;
  /** User email address */
  email: string;
  /** User display name */
  name?: string;
  /** User profile image */
  image?: string;
  /** OAuth scopes granted to the token */
  scopes: string[];
  /** OAuth client ID that requested the token */
  clientId: string;
  /** Token expiration timestamp (milliseconds since epoch) */
  expiresAt: number;
}

export class OAuthTokenValidator {
  private config: OAuthConfig;

  constructor(config: OAuthConfig) {
    this.config = config;
  }

  /**
   * Validate a Bearer token using HTTP request to mcp-obs platform
   * @param token The Bearer token to validate
   * @returns AuthContext if valid, null if invalid
   */
  async validateToken(token: string): Promise<AuthContext | null> {
    try {
      const platformUrl = this.config.platformUrl || `https://${this.config.serverSlug}.mcp-obs.com`;

      if (this.config.debug) {
        console.log(`[OAuth] Validating token with platform: ${platformUrl}`);
      }

      // Make HTTP request to introspect token
      const response = await fetch(`${platformUrl}/api/mcp-oauth/introspect`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: new URLSearchParams({
          token,
          server_slug: this.config.serverSlug
        })
      });

      if (!response.ok) {
        if (this.config.debug) {
          console.error(`[OAuth] HTTP error ${response.status}: ${response.statusText}`);
        }
        return null;
      }

      const introspection = await response.json();

      if (!introspection.active) {
        if (this.config.debug) {
          console.error('[OAuth] Token is not active');
        }
        return null;
      }

      const authContext: AuthContext = {
        userId: introspection.sub || introspection['mcp:user_id'] || 'unknown',
        email: introspection.username || 'unknown',
        name: introspection.name,
        image: introspection.picture,
        scopes: introspection.scope ? introspection.scope.split(' ') : [],
        clientId: introspection.client_id || 'unknown',
        expiresAt: introspection.exp ? introspection.exp * 1000 : Date.now() + 3600000
      };

      if (this.config.debug) {
        console.log(`[OAuth] Token validation successful for user: ${authContext.email}`);
      }

      return authContext;

    } catch (error) {
      console.error('[OAuth] Token validation error:', error);
      return null;
    }
  }

  /**
   * Extract Bearer token from Authorization header
   * @param authHeader The Authorization header value
   * @returns The Bearer token or null if not found
   */
  extractBearerToken(authHeader?: string): string | null {
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      return null;
    }
    return authHeader.slice(7);
  }

  /**
   * Check if token has required scopes
   * @param authContext The validated auth context
   * @param requiredScopes Array of required scopes
   * @returns True if all required scopes are present
   */
  hasRequiredScopes(authContext: AuthContext, requiredScopes: string[]): boolean {
    return requiredScopes.every(scope => authContext.scopes.includes(scope));
  }

}